daffodil-test/src/test/resources/org/apache/daffodil/section23/dfdl_expressions/expressions.tdml

```diff
@@ -7253,8 +7253,7 @@
     <tdml:document>1,2,3</tdml:document>
     <tdml:errors>
       <tdml:error>Schema Definition Error</tdml:error>
-      <tdml:error>Statically ambiguous or query-style paths not supported</tdml:error>
-      <tdml:error>ex:a</tdml:error>
+      <tdml:error>Path step 'ex:{http://example.com}a' ambiguous</tdml:error>
```

Comment:
Steve's hypothesis sounds worth checking to me.  Mike, I wonder if running "daffodil parse ..." by hand on the given TDML test case(s) with both old and new daffodil builds would show any changes in the diagnostics.  If you don't see any change, then the saving and reloading of the parser must indeed be causing the change in the diagnostics the TDML runner sees.

Reply:
I think it is this line here:

https://github.com/apache/daffodil/blob/main/daffodil-tdml-lib/src/main/scala/org/apache/daffodil/tdml/TDMLRunner.scala#L978

And I think what is happening is that compile-time warnings are getting converted to errors but runtime warnings are not.

Reply:
And I think the "change" is that there was a compile-time warning but also a runtime error for the same condition. So the tests were originally checking for the warning (which got converted to an error), but now that those are not serialized, we are now seeing the associated runtime error.

Reply:
Yeah, something like that is going on. I spent a little time trying to better understand our TDML runner, because it's still not totally clear to me. I think I have a better idea the core issue (I thin we have a couple related bugs).

In this particular case, the `checkDiagnosticMessages` function is called to verify all expected diagnostics are found:

https://github.com/apache/daffodil/blob/main/daffodil-tdml-lib/src/main/scala/org/apache/daffodil/tdml/TDMLRunner.scala#L928-L949

The `diagnostics` parameter contains a combination of diagnostics from schema compilation and from the parse . This function then calls `verifyAllDiagnosticsFound` twice, both times passing in those diagnostics and either the expected errors or expected warnings. The `verifyAllDiagnosticsFound` function iterates over the expected errors or warnings, and makes sure they exist in any of the passed in actual diagnostics.

https://github.com/apache/daffodil/blob/main/daffodil-tdml-lib/src/main/scala/org/apache/daffodil/tdml/TDMLRunner.scala#L1784-L1799

But that does no checks to make sure the type of the expected diagnostics matches the type of the actual diagnostics (i.e. error vs warning). So right now, I think there is no difference between `<tdml:errors>` and `<tdml:warnings>` as far as diagnostics checking goes. Note that there is a difference in the logic in other places, for example, if there are expected errors then we may require a failed parse result. So there are other cases where it matters. But this means if there is a parse error, any warnings can be `<tdml:error>` or `<tdml:warning>` and the warning diagnostic will be found.

So that's a bug that needs to be fixed. `verifyAllDiagnosticsFound` should look at the type of each `expectedDiag` and should should accept matches in `actualDiags` of the appropriate error/warning type. Whether or not that is fixed as part of the PR or a new one doesn't really matter to me, it is sort of a different issue. Though it's also probably a fairly small change, so might be worth including in this since it is somewhat related. Note that I think if we fixed this, the `tdml:error` in this test case would need to change to a `tdml:warning` and a new `tdml:error` would be added with this new error message.

But that doesn't explain why the error message changed. So what caused that? You're right that it's because we aren't serializing those warnings, so a save/reload loses those warnings, but the TDMLRuner should capture them and use them for later. When we get a successful `CompileResult`, it is a tuple of compilation warnings and the actual processor, so we do have those warnings even if the processor is saved/reloaded.

I think the issue is where we call `runParseExepctedErrors/Success` here:

https://github.com/apache/daffodil/blob/main/daffodil-tdml-lib/src/main/scala/org/apache/daffodil/tdml/TDMLRunner.scala#L976-L1012

In that above chunk, we save the processor from the CompileResult to the `processor` var, but we just ignore whatever diagnostics were part of the `CompileResult`. To get the compile diagnostics for later verification, we end up calling `processor.getDiagnostics` in a number of places. But if we saved/reloaded the processor, the processor no longer has those diagnostics.

So I think the issue is that we should never call `processor.getDiagnostics` (that is an unreliable way to get compile time diags due to this change), but we should instead pass around the diagnostics from the `CompileResult` and use that. And the `getDiagnostics` function should probably just be removed from the `TDMLDFDLProcessor` API--any diagnostics related to compilation should come from the `CompileResult`.

Reply:
I agree with @stevedlawrence's suggestion to remove `processor.getDiagnostics` calls and get diagnostics from the `CompileResult`, `TDMLParseResult`, and `TDMLUnparseResult` instead.  When I implemented the `TDMLDFDLProcessor` API in `DaffodilCTDMLDFDLProcessor`, I already made its `isError` and `getDiagnostics` no-ops:

```scala
  // We return errors and diagnostics in TDMLParseResult and TDMLUnparseResult
  override def isError: Boolean = false
  override def getDiagnostics: Seq[Diagnostic] = Seq.empty
```

These bigger changes may break some tests, but it should be worth fixing some tests.

Reply:
I've updated the runner to now check to see if warnings match up to warnings and if errors match up to errors. Quite a few existing tests had to change as they were relying on the previous behavior.  I'm looking into removing the `getDiagnostics` and `isError` definitions next.